### DISCUSSION TITLE
Retrieving upstream Task IDs within a mapped task

### DISCUSSION THREAD

**Discussion by mattmelgard** [apache/airflow#29826] [Q&A]
I'm trying to figure out how to get the `upstream_task_ids` from the Airflow context within a Dynamically Mapped Task and having some trouble doing so.

For some context (without getting too into the weeds here), I'm trying to instrument our Airflow DAGs with Datadog tracing and I created a decorator to do so. In order to pass the trace metadata between each task so that I can link the parent task's span to the child task's span in a way that works in tasks that use multi-processing, I'm passing the Datadog span and trace ID in XCOM variables where the key is the task ID and the value is an object containing the span and trace ID.

In order to gather the upstream task IDs in a mapped task (so that I can then retrieve the trace and span ID of the upstream task and connect to the original trace), I'm doing something like the following within the aforementioned decorator:

```
from airflow.operators.python import get_current_context

...
def trace_func(func):

    @functools.wraps(func)
    def traced_func(*args, **kwargs):
       ...
        airflow_context = get_current_context()
        current_task = airflow_context.get('task')
        upstream_task_ids = list(current_task.upstream_task_ids)
       ...
...
``` 

but I seem to be getting back an empty list for the `upstream_task_ids` when a mapped task is being executed. Is there some other way that I should be trying to access the upstream task IDs within a mapped task?

**Comment by potiuk** [comment]
Might be a missing feature of mapped tasks @uranusjr ?

**Comment by potiuk** [comment]
BTW.  I think that might be something interesting @ferruzzi  while implementing ~AIP-53 https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-53+OpenLineage+in+Airflow~ AIP-49 https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-49+OpenTelemetry+Support+for+Apache+Airflow

**Comment by ferruzzi** [comment]
I haven't fallen that deeply down the rabbit hole yet, but I'll see if I can come up with something and definitely be keeping an eye on this in case someone smarter than me chimes in first.  :+1:

**Comment by uranusjr** [comment]
IIRC we kind of sort of knew about this; the unmapped operator object (what you get in `context["task"]`) does not have many of the DAG structure attributes set because it can’t (it would break the actual DAG structure). For `upstream_task_ids` (and `downstream_task_ids`) it’s likely OK since those are just plain sets of strings, but I think we also were unsure whether those are supported public API either.

If you want to experiement, add a couple lines in `unmap` after the unmapped operator is created:

```python
op.task_id = self.task_id  # This line should already exist.
op.upstream_task_ids = self.upstream_task_ids
op.downstream_task_ids = self.downstream_task_ids
return op
```

**Reply by mattmelgard** [reply]
I'm not sure I follow, is this a solution you're suggesting here? Where would I need to be adding this? In the task itself?

**Reply by uranusjr** [reply]
Sorry, I was talking in the “if you want to modify Airflow itself” sense (mostly responding to @ferruzzi above), this is not something you can do in DAG code.

**Comment by jabbera** [comment]
@mattmelgard  did you ever solve this issue?
